File: src/taeeun/week1/거리두기 확인하기.js
function solution(places) {
    var answer = [];
    const dx=[0, 0, 1, -1];
    const dy=[-1, 1, 0, 0];
    
    const isReachable=(x, y)=>{
        if (x<0 || x>=5 || y<0 || y>=5) return false;
        return true;
    }
    
    const bfs=(place, x, y)=>{
        const visited=Array.from(Array(5), () => new Array(5).fill(false));
        let q=[[x,y,0]];
        visited[x][y]=true;
        while (q.length) {
            const [currentX, currentY, dist] = q.shift();
            if (dist==2) continue; //거리두기가 잘 지켜지고 있음
            for (let i=0; i<4; i++) {
                const nextX=currentX+dx[i];
                const nextY=currentY+dy[i];
                if (!isReachable(nextX, nextY)) continue; //접근 불가능
                if (visited[nextX][nextY] || place[nextX][nextY]=='X') continue;
                if (place[nextX][nextY]=='O') {
                    //빈 테이블이라면 일단 계속 확인
                    visited[nextX][nextY]=true;
                    q.push([nextX, nextY, dist+1]);
                }
                if (place[nextX][nextY]=='P') {
                    return false;
                }
            }
            
        }
        return true;
    }
    
    const findAnswer=(place)=>{
        for (let i=0; i<5; i++) {
            for (let j=0; j<5; j++) {
                if (place[i][j]=='P') {
                    if (bfs(place, i, j)==false) {
                        return 0;
                    }
                }
            }
        }  
        return 1;
    }
    
    for (let k=0; k<places.length; k++) {
        const place=places[k].map((e)=>e.split(''));
        answer.push(findAnswer(place));
    }
    
    return answer;
}

File: src/jimin/week1/거리두기 확인하기.py
from collections import deque

dx = [0,0,-1,1]
dy = [-1,1,0,0]

def solution(places):
    answer = []
    visited = [[0] * 5 for _ in range(5)]
    # P 상하좌우에 P -> 못지킴
    # P 상하좌우에 O -> 따라갔다가 상하좌우에 P -> 못지킴
    def bfs(y,x, visited):
        queue = deque()
        queue.append((y,x,0))
        while queue:
            y,x,depth = queue.popleft()
            visited[y][x] = 1
            if depth == 2:
                break
            for i in range(4):
                nx, ny = x + dx[i], y + dy[i]
                if 0 <= nx < 5 and 0 <= ny < 5 and not visited[ny][nx]:
                    if place[ny][nx] == 'P':
                        return False
                    if place[ny][nx] == 'O':
                        queue.append((ny,nx, depth+1))
        return True

    def dist(place):
        for i in range(5):
            for j in range(5):
                if place[i][j] == 'P':
                    if not bfs(i,j, visited):
                        return False
        return True

    for place in places:
        visited = [[0] * 5 for _ in range(5)]
        if not dist(place):
            answer.append(0)
        else:
            answer.append(1)
    return answer

places = [["PXPXP", "OPXPX", "PXPXP", "XPXPX", "PXPXP" ], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"], ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"], ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]]
print(solution(places))

# dx = [0,0,-1,1]
# dy = [-1,1,0,0]
# def solution(places):
#     answer = []
#     visited = [[0] * 5 for _ in range(5)]
#     # P 상하좌우에 P -> 못지킴
#     # P 상하좌우에 O -> 따라갔다가 상하좌우에 P -> 못지킴
#     def dfs(y,x, depth, place, visited):
#         visited[y][x] = 1
#         if depth == 2:
#             return True
#         for i in range(4):
#             nx, ny = x+dx[i], y+dy[i]
#             if 0 <= nx < 5 and 0 <= ny < 5 and not visited[ny][nx]:
#                 if place[ny][nx] == 'P':
#                     return False
#                 if place[ny][nx] == 'O':
#                     if not dfs(ny,nx,depth+1,place, visited):
#                         return False
#         return True
#
#     def dist(place):
#         for i in range(5):
#             for j in range(5):
#                 if place[i][j] == 'P':
#                     if not dfs(i,j,0,place, visited):
#                         return False
#         return True
#
#     for place in places:
#         visited = [[0] * 5 for _ in range(5)]
#         if not dist(place):
#             answer.append(0)
#         else:
#             answer.append(1)
#     return answer